src/Discord.Net.Rest/Net/DefaultRestClient.cs

```diff
@@ -101,62 +101,68 @@ public void SetCancelToken(CancellationToken cancelToken)
             IEnumerable<KeyValuePair<string, IEnumerable<string>>> requestHeaders = null)
         {
             string uri = Path.Combine(_baseUrl, endpoint);
-            using (var restRequest = new HttpRequestMessage(GetMethod(method), uri))
+
+            // HttpRequestMessage implements IDisposable but we do not need to dispose it as it merely disposes of its Content property,
+            // which we can do as needed. And regarding that, we do not want to take responsibility for disposing of content provided by
+            // the caller of this function, since it's possible that the caller wants to reuse it or is forced to reuse it because of a
+            // 429 response. Therefore, by convention, we only dispose the content objects created in this function (if any).
+            //
+            // See this comment explaining why this is safe: https://github.com/aspnet/Security/issues/886#issuecomment-229181249
+            // See also the source for HttpRequestMessage: https://github.com/microsoft/referencesource/blob/master/System/net/System/Net/Http/HttpRequestMessage.cs
+#pragma warning disable IDISP004
+            var restRequest = new HttpRequestMessage(GetMethod(method), uri);
+#pragma warning restore IDISP004
+
+            if (reason != null)
+                restRequest.Headers.Add("X-Audit-Log-Reason", Uri.EscapeDataString(reason));
+            if (requestHeaders != null)
+                foreach (var header in requestHeaders)
+                    restRequest.Headers.Add(header.Key, header.Value);
+            var content = new MultipartFormDataContent("Upload----" + DateTime.Now.ToString(CultureInfo.InvariantCulture));
+
+            static StreamContent GetStreamContent(Stream stream)
             {
-                if (reason != null)
-                    restRequest.Headers.Add("X-Audit-Log-Reason", Uri.EscapeDataString(reason));
-                if (requestHeaders != null)
-                    foreach (var header in requestHeaders)
-                        restRequest.Headers.Add(header.Key, header.Value);
-                var content = new MultipartFormDataContent("Upload----" + DateTime.Now.ToString(CultureInfo.InvariantCulture));
-                MemoryStream memoryStream = null;
-                if (multipartParams != null)
+                if (stream.CanSeek)
+                {
+                    // Reset back to the beginning; it may have been used elsewhere or in a previous request.
+                    stream.Position = 0;
```

Comment:
I debated between whether to seek back to the beginning before or after the request is processed. I decided to do it before because I think in general we can't assume that a stream has been reset to the beginning, and the integrating app shouldn't assume that about its own streams either.